File: implementations/python/src/contact_enrichment/infrastructure/security.py
"""Security kernel for MAC enforcement."""
import logging
from typing import Protocol

from contact_enrichment.domain.contact import IntegrityLevel, SecurityLabel
from contact_enrichment.infrastructure.repository import SecurityContext

logger = logging.getLogger(__name__)


class SecurityError(Exception):
    """Base exception for security violations."""

    pass


class AccessDeniedError(SecurityError):
    """Exception raised when access is denied."""

    pass


class MfaRequiredError(SecurityError):
    """Exception raised when MFA is required but not verified."""

    pass


class SecurityKernel(Protocol):
    """Security kernel interface for MAC enforcement."""

    def authorize_read(self, context: SecurityContext, data_label: SecurityLabel) -> None:
        """Authorize read access (Bell-LaPadula: no read up)."""
        ...

    def authorize_write(self, context: SecurityContext, data_label: SecurityLabel) -> None:
        """Authorize write access (Biba: no write down + Bell-LaPadula)."""
        ...

    def authorize_contact_creation(self, context: SecurityContext) -> None:
        """Authorize contact creation."""
        ...

    def authorize_enrichment(
        self,
        context: SecurityContext,
        contact_label: SecurityLabel,
        attribute_label: SecurityLabel,
    ) -> None:
        """Authorize enrichment operation."""
        ...


class TrustedSecurityKernel:
    """Trusted Security Kernel implementing Bell-LaPadula and Biba models.

    Implements:
    - Bell-LaPadula model (no read up)
    - Biba model (no write down)
    - Compartmentalization (need-to-know)
    - Audit logging for all decisions
    """

    def authorize_read(self, context: SecurityContext, data_label: SecurityLabel) -> None:
        """Authorize read access with Bell-LaPadula enforcement."""
        request_id = context.request_id

        logger.debug(
            f"Authorization check [{request_id}]: principal={context.principal_id}, "
            f"operation=READ, dataLabel={data_label}"
        )

        # Bell-LaPadula: No read up
        if not context.clearance.dominates(data_label):
            logger.warning(
                f"AUTHORIZATION DENIED [{request_id}]: Clearance insufficient for read - "
                f"principal={context.principal_id}, required={data_label}, "
                f"actual={context.clearance}"
            )
            raise AccessDeniedError(
                "Access denied: Your clearance does not dominate the data classification"
            )

        # Check compartments (need-to-know)
        if not context.clearance.compartments.issuperset(data_label.compartments):
            logger.warning(
                f"AUTHORIZATION DENIED [{request_id}]: Missing compartments - "
                f"principal={context.principal_id}, required={data_label.compartments}, "
                f"actual={context.clearance.compartments}"
            )
            raise AccessDeniedError(
                "Access denied: Missing required compartments (need-to-know)"
            )

        logger.info(
            f"AUTHORIZATION GRANTED [{request_id}]: principal={context.principal_id}, "
            f"operation=READ"
        )

    def authorize_write(self, context: SecurityContext, data_label: SecurityLabel) -> None:
        """Authorize write access with Biba and Bell-LaPadula enforcement."""
        request_id = context.request_id

        logger.debug(
            f"Authorization check [{request_id}]: principal={context.principal_id}, "
            f"operation=WRITE, dataLabel={data_label}"
        )

        # Biba: No write down (integrity)
        if data_label.integrity.value > context.clearance.integrity.value:
            logger.warning(
                f"AUTHORIZATION DENIED [{request_id}]: Integrity level insufficient - "
                f"principal={context.principal_id}, required={data_label.integrity}, "
                f"actual={context.clearance.integrity}"
            )
            raise AccessDeniedError(
                "Access denied: Cannot write to higher integrity level (no write down)"
            )

        # Must also be able to read (Bell-LaPadula)
        self.authorize_read(context, data_label)

        logger.info(
            f"AUTHORIZATION GRANTED [{request_id}]: principal={context.principal_id}, "
            f"operation=WRITE"
        )

    def authorize_contact_creation(self, context: SecurityContext) -> None:
        """Authorize contact creation with integrity and MFA checks."""
        request_id = context.request_id

        logger.debug(
            f"Authorization check [{request_id}]: principal={context.principal_id}, "
            f"operation=CREATE_CONTACT"
        )

        # Require minimum Medium integrity for contact creation
        if context.clearance.integrity.value < IntegrityLevel.MEDIUM.value:
            logger.warning(
                f"AUTHORIZATION DENIED [{request_id}]: Insufficient integrity for "
                f"contact creation - principal={context.principal_id}, "
                f"actual={context.clearance.integrity}"
            )
            raise AccessDeniedError(
                "Access denied: Minimum Medium integrity level required for contact creation"
            )

        # Require MFA for sensitive operations
        if not context.mfa_verified:
            logger.warning(
                f"AUTHORIZATION DENIED [{request_id}]: MFA required - "
                f"principal={context.principal_id}"
            )
            raise MfaRequiredError("Access denied: Multi-factor authentication required")

        # Require declared purpose for GDPR/CCPA compliance
        if not context.declared_purpose:
            logger.warning(
                f"AUTHORIZATION DENIED [{request_id}]: No processing purpose declared - "
                f"principal={context.principal_id}"
            )
            raise AccessDeniedError(
                "Access denied: Processing purpose must be declared (GDPR compliance)"
            )

        logger.info(
            f"AUTHORIZATION GRANTED [{request_id}]: principal={context.principal_id}, "
            f"operation=CREATE_CONTACT"
        )

    def authorize_enrichment(
        self,
        context: SecurityContext,
        contact_label: SecurityLabel,
        attribute_label: SecurityLabel,
    ) -> None:
        """Authorize enrichment with label validation."""
        request_id = context.request_id

        logger.debug(
            f"Authorization check [{request_id}]: principal={context.principal_id}, "
            f"operation=ENRICH, contactLabel={contact_label}, "
            f"attributeLabel={attribute_label}"
        )

        # Must be able to write to contact
        self.authorize_write(context, contact_label)

        # Attribute label must not exceed contact label (prevent upgrade)
        if not contact_label.dominates(attribute_label):
            logger.warning(
                f"AUTHORIZATION DENIED [{request_id}]: Attribute label exceeds contact "
                f"label - principal={context.principal_id}, contactLabel={contact_label}, "
                f"attributeLabel={attribute_label}"
            )
            raise AccessDeniedError(
                "Access denied: Attribute security label exceeds contact label"
            )

        logger.info(
            f"AUTHORIZATION GRANTED [{request_id}]: principal={context.principal_id}, "
            f"operation=ENRICH"
        )

File: implementations/rust/src/infrastructure/security.rs
// Security kernel for MAC enforcement
use crate::domain::contact::SecurityLabel;
use crate::infrastructure::repository::SecurityContext;

/// Security Kernel - Trusted Computing Base (TCB) for MAC enforcement.
///
/// Implements:
/// - Bell-LaPadula model (no read up)
/// - Biba model (no write down)
/// - Compartmentalization (need-to-know)
/// - Audit logging for all decisions
pub trait SecurityKernel: Send + Sync {
    /// Authorize read access (Bell-LaPadula: no read up).
    fn authorize_read(
        &self,
        context: &SecurityContext,
        data_label: &SecurityLabel,
    ) -> Result<(), SecurityError>;

    /// Authorize write access (Biba: no write down + Bell-LaPadula).
    fn authorize_write(
        &self,
        context: &SecurityContext,
        data_label: &SecurityLabel,
    ) -> Result<(), SecurityError>;

    /// Authorize contact creation.
    fn authorize_contact_creation(&self, context: &SecurityContext) -> Result<(), SecurityError>;

    /// Authorize enrichment operation.
    fn authorize_enrichment(
        &self,
        context: &SecurityContext,
        contact_label: &SecurityLabel,
        attribute_label: &SecurityLabel,
    ) -> Result<(), SecurityError>;
}

/// Trusted security kernel implementation.
pub struct TrustedSecurityKernel;

impl TrustedSecurityKernel {
    pub fn new() -> Self {
        Self
    }
}

impl Default for TrustedSecurityKernel {
    fn default() -> Self {
        Self::new()
    }
}

impl SecurityKernel for TrustedSecurityKernel {
    fn authorize_read(
        &self,
        context: &SecurityContext,
        data_label: &SecurityLabel,
    ) -> Result<(), SecurityError> {
        // Bell-LaPadula: No read up
        // Human note: clearance must dominate data label (level + compartments). This prevents
        // accidental data leaks from higher classifications to lower-cleared principals.
        if !context.clearance.dominates(data_label) {
            tracing::warn!(
                principal_id = %context.principal_id,
                required_label = ?data_label,
                actual_clearance = ?context.clearance,
                "Authorization denied: insufficient clearance for read"
            );

            return Err(SecurityError::AccessDenied(
                "Your clearance does not dominate the data classification".to_string(),
            ));
        }

        // Check compartments (need-to-know)
        if !context
            .clearance
            .compartments()
            .is_superset(data_label.compartments())
        {
            tracing::warn!(
                principal_id = %context.principal_id,
required_compartments = ?data_label.compartments(),
                actual_compartments = ?context.clearance.compartments(),
                "Authorization denied: missing compartments"
            );

            return Err(SecurityError::AccessDenied(
                "Missing required compartments (need-to-know)".to_string(),
            ));
        }

        tracing::info!(
            principal_id = %context.principal_id,
            operation = "READ",
            "Authorization granted"
        );

        Ok(())
    }

    fn authorize_write(
        &self,
        context: &SecurityContext,
        data_label: &SecurityLabel,
    ) -> Result<(), SecurityError> {
        // Biba: No write down (integrity)
        if data_label.integrity() > context.clearance.integrity() {
            tracing::warn!(
                principal_id = %context.principal_id,
                required_integrity = ?data_label.integrity(),
                actual_integrity = ?context.clearance.integrity(),
                "Authorization denied: insufficient integrity for write"
            );

            return Err(SecurityError::AccessDenied(
                "Cannot write to higher integrity level (no write down)".to_string(),
            ));
        }

        // Must also be able to read (Bell-LaPadula)
        self.authorize_read(context, data_label)?;

        tracing::info!(
            principal_id = %context.principal_id,
            operation = "WRITE",
            "Authorization granted"
        );

        Ok(())
    }

    fn authorize_contact_creation(&self, context: &SecurityContext) -> Result<(), SecurityError> {
        use crate::domain::contact::IntegrityLevel;

        // Require minimum Medium integrity for contact creation
        if context.clearance.integrity() < IntegrityLevel::Medium {
            tracing::warn!(
                principal_id = %context.principal_id,
                actual_integrity = ?context.clearance.integrity(),
                "Authorization denied: insufficient integrity for contact creation"
            );

            return Err(SecurityError::AccessDenied(
                "Minimum Medium integrity level required for contact creation".to_string(),
            ));
        }

        // Require MFA for sensitive operations
        if !context.mfa_verified {
            tracing::warn!(
                principal_id = %context.principal_id,
                "Authorization denied: MFA required"
            );

            return Err(SecurityError::MfaRequired);
        }

        // Require declared purpose for GDPR/CCPA compliance
        if context.declared_purpose.is_none() {
            tracing::warn!(
                principal_id = %context.principal_id,
                "Authorization denied: no processing purpose declared"
            );

            return Err(SecurityError::AccessDenied(
                "Processing purpose must be declared (GDPR compliance)".to_string(),
            ));
        }

        tracing::info!(
            principal_id = %context.principal_id,
            operation = "CREATE_CONTACT",
            "Authorization granted"
        );

        Ok(())
    }

    fn authorize_enrichment(
        &self,
        context: &SecurityContext,
        contact_label: &SecurityLabel,
        attribute_label: &SecurityLabel,
    ) -> Result<(), SecurityError> {
        // Must be able to write to contact
        self.authorize_write(context, contact_label)?;

        // Attribute label must not exceed contact label (prevent upgrade)
        if !contact_label.dominates(attribute_label) {
            tracing::warn!(
                principal_id = %context.principal_id,
                contact_label = ?contact_label,
                attribute_label = ?attribute_label,
                "Authorization denied: attribute label exceeds contact label"
            );

            return Err(SecurityError::AccessDenied(
                "Attribute security label exceeds contact label".to_string(),
            ));
        }

        tracing::info!(
            principal_id = %context.principal_id,
            operation = "ENRICH",
            "Authorization granted"
        );

        Ok(())
    }
}

/// Security errors.
#[derive(Debug, thiserror::Error)]
pub enum SecurityError {
    #[error("Access denied: {0}")]
    AccessDenied(String),

    #[error("Multi-factor authentication required")]
    MfaRequired,

    #[error("Session expired")]
    SessionExpired,
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::domain::contact::{ConfidentialityLevel, IntegrityLevel};
    use std::collections::HashSet;
    use uuid::Uuid;

    #[test]
    fn test_authorize_read_with_sufficient_clearance() {
        let kernel = TrustedSecurityKernel::new();

        let context = SecurityContext {
            request_id: Uuid::new_v4(),
            principal_id: Uuid::new_v4(),
            clearance: SecurityLabel::new(
                ConfidentialityLevel::Confidential,
                IntegrityLevel::High,
                vec!["PII".to_string()],
            ),
            mfa_verified: true,
            declared_purpose: Some("testing".to_string()),
        };

        let data_label = SecurityLabel::new(
            ConfidentialityLevel::Internal,
            IntegrityLevel::Medium,
            vec![]
        );

        assert!(kernel.authorize_read(&context, &data_label).is_ok());
    }

    #[test]
    fn test_authorize_read_fails_with_insufficient_clearance() {
        let kernel = TrustedSecurityKernel::new();

        let context = SecurityContext {
            request_id: Uuid::new_v4(),
            principal_id: Uuid::new_v4(),
            clearance: SecurityLabel::new(
                ConfidentialityLevel::Internal,
                IntegrityLevel::Medium,
                vec![]
            ),
            mfa_verified: true,
            declared_purpose: Some("testing".to_string()),
        };

        let data_label = SecurityLabel::new(
            ConfidentialityLevel::Confidential,
            IntegrityLevel::High,
            vec!["PII".to_string()]
        );

        assert!(kernel.authorize_read(&context, &data_label).is_err());
    }
}
